src/impl/transact_log_handler.cpp

```diff
@@ -267,12 +279,18 @@ class TransactLogObserver : public TransactLogValidationMixin, public MarkDirtyM
             if (context && old_version != sg.get_version_of_current_transaction()) {
                 context->did_change({}, {});
             }
+            m_notifiers.deliver(sg);
+            m_notifiers.after_advance();
             return;
         }
 
         std::sort(begin(m_observers), end(m_observers));
         func(*this);
-        context->did_change(m_observers, invalidated);
+        if (context)
+            context->did_change(m_observers, invalidated, old_version != sg.get_version_of_current_transaction());
+        m_notifiers.package_and_wait(sg.get_version_of_current_transaction().version); // is a no-op if parse_complete() was called
```

Comment:
If `parse_complete()` is not called it looks like we'll call `deliver` / `after_advance` without first calling `begin_advance`. Is that ok?